### PR TITLE
Switch to YAML config files

### PR DESCRIPTION
**Pull request by mavam**
- [x] Rebase onto master after #1055 is merged.
- [x] Rebase onto master after #1067 is merged.

### CODE REVIEW

**Comment by mavam** [comment]
Note to the reviewer: this PR is best reviewed commit by commit.

File: vast.conf.example [hunk 1]
```diff
@@ -1,447 +1,411 @@
...
+# The `vast pivot` command extracts related events of a given type.
+pivot:
+  # The output format.
+  # format: "json"
+
+  # For additionally available options, see export.pcap.
+
+# The `vast status` command prints a JSON-formatted status summary of the node.
+status:
+  # No further configuration options are available. The system options apply.
```
Comment (dominiklohmann) [inline vast.conf.example:261]: This is not true.

File: vast.conf.example [hunk 2]
```diff
@@ -1,447 +1,411 @@
...
+
+  # Excludes listed components from logging.
+  component-blacklist:
+    - caf
+    - caf_flow
+    - caf_stream
+
+#scheduler:
+  # Accepted alternative: "sharing".
+  #policy: "stealing"
```
Comment (dominiklohmann) [inline vast.conf.example:363]:
```suggestion
  #policy: stealing
```

File: vast.conf.example [hunk 3]
```diff
@@ -1,447 +1,411 @@
...
+
+  # Configures whether the scheduler generates profiling output.
+  #enable-profiling: false
+
+  # Output file for profiler data (only if profiling is enabled).
+  #profiling-output-file: "/dev/null"
```
Comment (dominiklohmann) [inline vast.conf.example:369]:
```suggestion
  #profiling-output-file: /dev/null
```

File: libvast/src/system/configuration.cpp [hunk 4]
```diff
@@ -79,32 +83,99 @@ caf::error configuration::parse(int argc, char** argv) {
...
+    if (exists(config)) {
+      auto contents = load_contents(config);
+      if (!contents)
+        return contents.error();
+      auto yaml = from_yaml(*contents);
+      if (!yaml)
+        return yaml.error();
+      auto rec = caf::get_if<record>(&*yaml);
+      if (!rec)
+        return caf::make_error(ec::parse_error, "config file not a YAML map");
```
Comment (dominiklohmann) [inline libvast/src/system/configuration.cpp:109]: Please improve this error context. From a user's point of view, this does not currently make any sense.

File: libvast/src/system/configuration.cpp [hunk 5]
```diff
@@ -79,32 +83,99 @@ caf::error configuration::parse(int argc, char** argv) {
...
+      auto flat_yaml = flatten(*rec);
+      // Erase all null values because a caf::config_value has no such notion.
+      for (auto i = flat_yaml.begin(); i != flat_yaml.end();) {
+        if (caf::holds_alternative<caf::none_t>(i->second))
+          i = flat_yaml.erase(i);
+        else
+          ++i;
+      }
```
Comment (dominiklohmann) [inline libvast/src/system/configuration.cpp:117]: This is faulty behavior w.r.t. the merging behavior below. `null` in a higher-precedence configuration file should override the existing setting.

File: libvast/src/system/configuration.cpp [hunk 6]
```diff
@@ -79,32 +83,99 @@ caf::error configuration::parse(int argc, char** argv) {
...
+                         [](const std::string& x) { return x; });
+    auto str = caf::visit(no_quote_stringify, val);
+    auto result = opt.parse(str);
+    if (!result) {
+      // We now try to parse strings as atom using a regex, since we get
+      // recursive types like lists for free this way. A string-vs-atom type
+      // clash is the only instance we currently cannot distinguish. Everything
+      // else is a true type clash.
+      // (With CAF 0.18, this heuristic will be obsolete.)
+      str = std::regex_replace(str, std::regex("\""), "'");
```
Comment (dominiklohmann) [inline libvast/src/system/configuration.cpp:145]: Please rewrite this in terms of `std::transform_if` for three reasons:
1. It is absolutely overcomplicating the code for readers and reviewers.
2. It is unnecessarily slow at runtime, especially the unnecessarily repeated regular expression compilation (these should _always_ be cached).
3. The inclusion of the regex header adds a lot of compile time overhead.

**Review by dominiklohmann — CHANGES_REQUESTED** [review]
I only looked at the code for now, will do the testing locally for a later review round.

Please rename the config file in this PR already and warn if a `vast.conf` file is found to reduce friction for the user.

File: vast.conf.example (hunk 1 above)
Comment (mavam) [reply vast.conf.example:261]: I just took this from the default. How about we change this as we are restructuring the YAML file and streamline the sections? Then it becomes also easier to see how the transformation from CAF to YAML config took place.

File: libvast/src/system/configuration.cpp (hunk 5 above)
Comment (mavam) [reply libvast/src/system/configuration.cpp:117]: So then we need to merge all `data` instances first.

File: libvast/src/system/configuration.cpp (hunk 6 above)
Comment (mavam) [reply libvast/src/system/configuration.cpp:145]: Turns out we had `detail::replace_all` already! 🙈

File: libvast/src/system/configuration.cpp (hunk 4 above)
Comment (mavam) [reply libvast/src/system/configuration.cpp:108]: Better?

**Comment by mavam** [comment]
CI currently fails due the issue to be fixed in #1067; marking this PR as *blocked* until merged.

File: vast.conf.example (hunk 1 above)
Comment (dominiklohmann) [reply vast.conf.example]: Alrighty.

**Review by dominiklohmann — APPROVED** [review]
This is working nicely in practice. I've manually verified the merging logic to work.

**Pull request merged**